pkg/provider/loadbalancer.go

```diff
@@ -0,0 +1,179 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/klog/v2"
+)
+
+type loadBalancer struct {
+	namespace  string
+	kubeClient kubernetes.Interface
+}
+
+var _ cloudprovider.LoadBalancer = &loadBalancer{}
+
+func (l *loadBalancer) GetLoadBalancer(ctx context.Context, clusterName string, service *corev1.Service) (status *corev1.LoadBalancerStatus, exists bool, err error) {
+	lbServiceName := l.GetLoadBalancerName(ctx, clusterName, service)
+	lbService, err := l.kubeClient.CoreV1().Services(l.namespace).Get(ctx, lbServiceName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			lbService = nil
+		} else {
+			return nil, false, fmt.Errorf("get LB service: %s", err)
+		}
+	}
+
+	if lbService == nil {
+		return nil, false, nil
+	} else {
+		return &lbService.Status.LoadBalancer, true, nil
+	}
+}
+
+func (l *loadBalancer) GetLoadBalancerName(ctx context.Context, clusterName string, service *corev1.Service) string {
+	return cloudprovider.DefaultLoadBalancerName(service)
+}
+
+func (l *loadBalancer) EnsureLoadBalancer(ctx context.Context, clusterName string, service *corev1.Service, nodes []*corev1.Node) (*corev1.LoadBalancerStatus, error) {
+	lbServiceName := l.GetLoadBalancerName(ctx, clusterName, service)
+	lbService, err := l.kubeClient.CoreV1().Services(l.namespace).Get(ctx, lbServiceName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			lbService = nil
+		} else {
+			return nil, fmt.Errorf("get LB service: %s", err)
+		}
+	}
+
+	if lbService == nil {
+		lbService := &corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      lbServiceName,
+				Namespace: l.namespace,
+			},
+			Spec: corev1.ServiceSpec{
+				Selector: map[string]string{
+					"cluster.x-k8s.io/cluster-name": clusterName,
+				},
+				Type:                  corev1.ServiceTypeLoadBalancer,
+				ExternalTrafficPolicy: service.Spec.ExternalTrafficPolicy,
+			},
+		}
+		setLBServicePorts(lbService, service)
+
+		if len(service.Spec.ExternalIPs) > 0 {
+			lbService.Spec.ExternalIPs = service.Spec.ExternalIPs
+		}
+		if service.Spec.LoadBalancerIP != "" {
+			lbService.Spec.LoadBalancerIP = service.Spec.LoadBalancerIP
+		}
+		if service.Spec.HealthCheckNodePort > 0 {
+			lbService.Spec.HealthCheckNodePort = service.Spec.HealthCheckNodePort
+		}
+
+		if _, err = l.kubeClient.CoreV1().Services(l.namespace).Create(ctx, lbService, metav1.CreateOptions{}); err != nil {
+			return nil, fmt.Errorf("create LB service: %s", err)
+		}
+	} else {
+		setLBServicePorts(lbService, service)
+		if _, err := l.kubeClient.CoreV1().Services(l.namespace).Update(ctx, lbService, metav1.UpdateOptions{}); err != nil {
+			return nil, fmt.Errorf("update LB service: %s", err)
+		}
+	}
+
+	if err := wait.PollUntil(5*time.Second, func() (bool, error) {
```

Comment:
without timeout, the reconcile loop will hang?

Reply:
Yes, this is what ensure wants